drgn_tools/task.py

```diff
@@ -197,6 +197,91 @@ def get_command(task: Object) -> str:
     return escape_ascii_string(task.comm.string_())
 
 
+def get_ppid(task: Object) -> int:
+    """
+    :returns: Parent PID of the task
+    """
+    return task.parent.pid.value_()
+
+
+def get_rss(prog: Program, task: Object) -> int:
+    """
+    :returns RSS value
+    """
+    try:
+        filerss = task.mm.rss_stat.count[0].counter.value_()
+        if filerss < 0:
+            filerss = 0
+        anonrss = task.mm.rss_stat.count[1].counter.value_()
+        if anonrss < 0:
+            anonrss = 0
+        shmrss = task.mm.rss_stat.count[3].counter.value_()
+        if shmrss < 0:
+            shmrss = 0
+        rss_mm = filerss + anonrss
+
+    except Exception:
+        rss_mm = 0
+
+    rss_task = 0
+    gid = task.tgid.value_()
+    for gtask in for_each_task(prog):
+        if gid == gtask.tgid.value_():
+            task_filerss = task_anonrss = task_shmrss = 0
+
+            try:
+                task_filerss = gtask.rss_stat.count[0].value_()
+                if task_filerss < 0:
+                    task_filerss = 0
+
+                task_anonrss = gtask.rss_stat.count[1].value_()
+                if task_anonrss < 0:
+                    task_anonrss = 0
+
+                task_shmrss = gtask.rss_stat.count[3].value_()
+                if task_shmrss < 0:
+                    task_shmrss = 0
+
+                rss_task += task_filerss + task_anonrss
+
+            except Exception:
+                rss_task = 0
```

Comment:
Multiple questions:

1. Iterating over every task in the system, just to find tasks of the same thread group, is inefficient. There is a field in `task_struct` called `thread_group` which appears to be a linked list of the threads. Can you use that instead?
2. Even then, iterating over every task in a thread group (i.e. process) could be quite expensive for processes with tons of threads. I think it might be worth adding a third argument to `get_rss()`, which would be a dict mapping `tgid` to `rss_task`. You would check this dictionary before computing it, and after computing a new value, you'd store it in the cache. This would ensure that you don't end up in an `O(N^2)` situation where you have a task with 1000 threads, and each time you do `get_rss()` on a thread you redo this, resulting in 1,000,000 iterations of the loop altogether.
3. Finally, what is the `except Exception` for? Which errors did you catch and why? Silently catching exceptions and continuing is not good for maintainable software. If there's a specific error that happens, it makes sense to catch that, but a blanked `except Exception` is a very wide net to cast, and will allow bugs to slip through.

Something like this maybe:
```suggestion
    gid = task.tgid.value_()
    if gid in rss_task_cache:
        rss_task = rss_task_cache[gid]
    else:
        rss_task = 0
        for gtask in list_for_each_entry("struct task_struct", 
                                                                task.thread_group.address_of_(),
                                                                "thread_group"):
            # note: this iteration doesn't include "task" itself, probably
            # need to add code to handle that.
            task_filerss = gtask.rss_stat.count[0].value_()
            if task_filerss < 0:
                task_filerss = 0

            task_anonrss = gtask.rss_stat.count[1].value_()
            if task_anonrss < 0:
                task_anonrss = 0

            task_shmrss = gtask.rss_stat.count[3].value_()
            if task_shmrss < 0:
                task_shmrss = 0

            rss_task += task_filerss + task_anonrss
        rss_task_cache[gid] = rss_task
```

Reply:
Agree with this and need a little modification ,

Reply:
> Let create set gid as index while traversing "for each task"
> and get the values directly from index when needed

I think this is what I'm proposing? `rss_task_cache` is a dictionary indexed by `gid`. The first time a task group is encountered, the task rss is computed and stored, and it will be reused for each subsequent thread of the task group.

The idea is that the caller will create that dictionary and the cache will be used for each call to avoid recomputing the task rss value.

We certainly don't want to do this in the `for_each_task()` function because most callers don't need memory usage information. But maybe I don't understand what you're suggesting. If I'm misunderstanding, I'm happy to take a look at a pseudocode or mock-up of what you're proposing!